### DISCUSSION TITLE
API for generating a Care Plan.

### DISCUSSION THREAD

**Discussion by gcangioli** [hl7-eu/pcsp#31] [Ideas]
Identify a solution to enable the automatic retrieval of the generated Care Plan

**Comment by gcangioli** [comment]
@davisart @StefanBeyer

**Reply by StefanBeyer** [reply]
Davide and I discussed this today and we came up with this solution:

To trigger the generation, a FHIR operation should be introduced. This FHIR operation immediately returns a link to the page where the generation can be triggered (by the click of a button). Technically, an administrator user will be used to access this link, so our end users are not required to manually enter credentials before they are able to trigger the generation.

For now, returning a link to the UI is the only viable option, as this does not require extensive development work on the platform and can be finished before the end of the project. In the future, an automatic generation might also be possible (using the same / a similar FHIR operation).

As we do not require the data to be stored on the platform (and do not want to because of legal reasons), all of the patients data should be deleted after the generation of a care plan. This might make troubleshooting a little bit harder, but as we can always resend the data, this should not really be an issue.

@gcangioli what are your thoughts about this? If this seems fine from your side, maybe you could propose a design for this operation, to be integrated into the implementation guide.

**Reply by costateixeira** [reply]
doesn't the $apply operation cover this?
https://build.fhir.org/plandefinition-operation-apply.html

**Reply by gcangioli** [reply]
Hi @StefanBeyer and @davisart ,

I'd be grateful if you can help me to understand what you'd like to achieve :-)

My understanding of the process was:
1. AT sends a minimal complete TS
2. The PCSP platform cache temporarily these data 
3. AT obtains the draft Care Plan

Solutions discussed were

(A) plain HL7 FHIR API
1. AT sends the TS (Single transaction ?)
2. AT asks for a CarePlan for that patient  (this request triggers the platform to generate the draft Care plan using the data that know about this patient)  ('plain' FHIR API ?)
3. Data of this patient are removed from the 'cache'

(B) ad hoc operation 
1. A new operation is defined using as input a Bundle and having as output  a CarePlan created based on the Bundle data 
2. data received using this operation are not stored in the server (unless explicitly requested)

not clear to me why you need to use an FHIR operation to return a web page with a button that the user need to click... 

Thanks for explaining

**Reply by StefanBeyer** [reply]
Hey @gcangioli,

yes so ideally we would only use the FHIR API to achieve this (either option A or B would be fine from our side).

However, @davisart noted that this would require too much work on their side (at least for now), as the care plan generation can currently only be triggered from the PCSP application server (using this button). As far as I understood the FHIR server is currently unable to invoke the application server, which seems to be the main issue here. Both solutions you mentioned would require a (synchronous) call from the FHIR server to the application server. Maybe @davisart could elaborate further.

I would not be unhappy if we found a way around that, but for the moment the solution @davisart proposed is acceptable to us.

@costateixeira this would absolutely be an option from my point of view (for solution A @gcangioli mentioned). The only issue might be, that as far as I'm aware, we do not have a corresponding plan definition instance for the care plan, but correct me if I am wrong.

**Reply by gcangioli** [reply]
> Hey @gcangioli,
> 
> yes so ideally we would only use the FHIR API to achieve this (either option A or B would be fine from our side).
> 
> However, @davisart noted that this would require too much work on their side (at least for now), as the care plan generation can currently only be triggered from the PCSP application server (using this button). As far as I understood the FHIR server is currently unable to invoke the application server, which seems to be the main issue here. Both solutions you mentioned would require a (synchronous) call from the FHIR server to the application server. Maybe @davisart could elaborate further.
> 
> I would not be unhappy if we found a way around that, but for the moment the solution @davisart proposed is acceptable to us.

If a human action is any case required I don't see the need of this FHIR operation :-)
@davisart ?

**Reply by gcangioli** [reply]
> @costateixeira this would absolutely be an option from my point of view (for solution A @gcangioli mentioned). The only issue might be, that as far as I'm aware, we do not have a corresponding plan definition instance for the care plan, but correct me if I am wrong.

Yes for the time being we have no PlanDefinition defined.

**Reply by StefanBeyer** [reply]
> > Hey @gcangioli,
> > yes so ideally we would only use the FHIR API to achieve this (either option A or B would be fine from our side).
> > However, @davisart noted that this would require too much work on their side (at least for now), as the care plan generation can currently only be triggered from the PCSP application server (using this button). As far as I understood the FHIR server is currently unable to invoke the application server, which seems to be the main issue here. Both solutions you mentioned would require a (synchronous) call from the FHIR server to the application server. Maybe @davisart could elaborate further.
> > I would not be unhappy if we found a way around that, but for the moment the solution @davisart proposed is acceptable to us.
> 
> If a human action is any case required I don't see the need of this FHIR operation :-) @davisart ?

If the returned URL is stable and can be derived from the information we already have (or can retrieve somehwere via FHIR) I agree. In this case we could simply generate it, instead of requesting it via an operation :-)

**Reply by StefanBeyer** [reply]
@davisart any updates on the URL we need to call? If we can somehow generate it with the information we already have, I think we can close this discussion.

**Reply by davisart** [reply]
@StefanBeyer yes, you can generate a Care Plan by call the button "Send" in Care Plan form. If you prefer as the system has been set we can have a call to look into and see if can work as expected.

**Reply by StefanBeyer** [reply]
@davisart @gcangioli from my experience with the current proprietary API I think it would be great to have an asynchronous extended FHIR operation (https://www.hl7.org/fhir/async.html) defined in the IG (e.g. on the Patient or CarePlan resource) to trigger the generation and to define a way to retrieve the status of the care plan generation, similar to the bulk export operation: https://hl7.org/fhir/uv/bulkdata/export.html#sequence-overview 

As the FHIR Asynchronous API still seems to be in a rather early stage and not yet supported by the currently used HAPI FHIR server (which only seems to support synchronous extended operations: https://hapifhir.io/hapi-fhir/docs/server_plain/rest_operations_operations.html), I guess it might not be feasable now to implement this, but maybe in a later version?
